### PR TITLE
test: use non-default base-id in run_envoy_test.sh

### PR DESCRIPTION
**Pull request by zuercher**
Commit Message: test: use non-default base-id in run_envoy_test.sh

Coverage has failed with an error related to using the default base-id 0.
Likely because two processes started Envoy with the same base-id. Use
the shell's PID or bazels TEST_RANDOM_SEED to avoid this situation.

Risk Level: low
Testing: fixes test
Docs Changes: n/a
Release Notes: n/a

Signed-off-by: Stephan Zuercher <zuercher@gmail.com>

### CODE REVIEW

**Comment by zuercher** [comment]
For reference the test failure was:
```
TEST: Launching ***** without --config-path or --config-yaml fails.
/build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/sandbox/processwrapper-sandbox/4971/execroot/*****/bazel-out/k8-fastbuild/bin/test/coverage/coverage_tests.runfiles/*****/source/exe/*****-static >& /build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/sandbox/processwrapper-sandbox/4971/execroot/*****/_tmp/54ce2028989fd7457dffae83b6c84c4f/*****.log
Turning perftools heap leak checking off
[2020-04-30 13:50:54.908][35710][debug][init] init manager Server destroyed
unable to bind domain socket with id=0 (see --base-id option)
     check [ 1 -eq 1 ] ...
     check grep At least one of --config-path or --config-yaml or Options::configProto() should be non-empty /build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/sandbox/processwrapper-sandbox/4971/execroot/*****/_tmp/54ce2028989fd7457dffae83b6c84c4f/*****.log ...
/build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/sandbox/processwrapper-sandbox/4971/execroot/*****/bazel-out/k8-fastbuild/bin/test/coverage/coverage_tests.runfiles/*****/test/integration/test_utility.sh:17: check
test/integration/run_*****_test.sh:15: expect_fail_with_error
test/integration/run_*****_test.sh:24: main
     failure at line 15
in 'Launching ***** without --config-path or --config-yaml fails.'
FAIL.
Failed bash -c "PYTHONPATH=$(dirname test/integration/run_*****_test.sh) test/integration/run_*****_test.sh "
```

**Comment by lizan** [comment]
@zuercher qq: I don't think I've seen this error before, does it fail the test or it just pass but results less coverage? if it pass I think we should make it fail?

**Comment by zuercher** [comment]
It happens when two Envoy processes create hot restart sockets with the same base id and the processes are running at the same time. Some of our tests run envoy-static and if two of these run concurrently with the same base-id one of them should report this error (even outside coverage).

That said,
* `test/integration/hot_restart_test.sh` uses base id 1.
* `test/extensions/filters/network/kafka/broker/integration_test/kafka_broker_integration_test.py`  uses a random number between 1 and 999999 (which I'll fix to use PID as well).
* `test/exe/version_out_test.sh` uses base-id 0, but should exit before the socket is created.

So it's not clear to me which test could have interfered. I don't see another test that actually opens the sockets (integration tests use the nop impl, hot restart tests mock out os sys calls, etc.)

**Comment by zuercher** [comment]
So two questions:

1. Is it possible coverage ran that test in two different shards? 
2. Maybe these tests that start envoy processes aren't really useful in the context of coverage and we should exclude them anyway?

I think I should still modify the test scripts to choose base ids that are not overlapping.

**Comment by zuercher** [comment]
Going to leave the kafka test. It's only run manually because it requires zookeeper and kafka to be running.

**Comment by zuercher** [comment]
And to answer your question, when his occurs, the build fails because Envoy exits before it reports the error about missing config.

File: test/integration/hotrestart_test.sh [hunk 1]
```diff
@@ -66,7 +66,7 @@ JSON_TEST_ARRAY+=("${HOT_RESTART_JSON_REUSE_PORT}")
 
 # Enable this test to work with --runs_per_test
 if [[ -z "${TEST_RANDOM_SEED}" ]]; then
-  BASE_ID=1
+  BASE_ID=$$
```
Comment (lizan) [inline test/integration/hotrestart_test.sh:69]: shall we always use the shell pid? it works even in runs_per_test right?

Comment (lizan) [reply test/integration/hotrestart_test.sh:69]: ping?

Comment (zuercher) [reply test/integration/hotrestart_test.sh:69]: Sorry, got distracted by work. I'll double check, but yes I think it's safe to just use the PID always.

Comment (zuercher) [reply test/integration/hotrestart_test.sh:69]: So, it appears that $$ when is always 16. I guess that's probably true for coverage as well.I think TEST_SHARD_INDEX should be the fallback there. I'll up date this to do that.

Comment (zuercher) [reply test/integration/hotrestart_test.sh]: And even better is that TEST_RANDOM_SEED is repeated for each test, so if you do `--runs_per_test=5` with multiple tests, they still collide.

**Comment by zuercher** [comment]
@lizan I think this is probably as safe as I can make it for now.

**Comment by mattklein123** [comment]
Do we still need this with the coverage binary split and sandboxing? If so would be nice to get this merged. Needs a master merge.

/wait

**Comment by alyssawilk** [comment]
I think it's worth landing, or disabling hot restart on tests that don't need it.  Any test run with --runs-per-test is subject to the dreaded "unable to bind domain socket" failure

**Comment by mattklein123** [comment]
OK let's definitely land. @zuercher can you update? @alyssawilk also can you point @zuercher at that line in your recent PR that should be reverted also I think?

**Comment by alyssawilk** [comment]
Yep, he can remove the options.setHotRestartDisabled(true); and comment
in test/config_test/config_test.cc

Addition here:
https://github.com/envoyproxy/envoy/pull/11066/files#diff-cee698f5ff6e31cd2ec9d1f978efa79d

**Comment by zuercher** [comment]
Working on getting an update out.

File: test/exe/main_common_test.cc [hunk 2]
```diff
@@ -38,34 +38,9 @@ class MainCommonTest : public testing::TestWithParam<Network::Address::IpVersion
...
-    // remainder to the random number.
-    const uint32_t four_digit_prime = 7919;
-#ifdef WIN32
-    return ::GetCurrentProcessId() * four_digit_prime +
-           random_generator_.random() % four_digit_prime;
-#else
-    return getpid() * four_digit_prime + random_generator_.random() % four_digit_prime;
-#endif
-  }
+        base_id_(TestEnvironment::chooseBaseId(3000000)),
```
Comment (mattklein123) [inline test/exe/main_common_test.cc:41]: From my quick read of this you have selected a range of base IDs in the 1M, 2M, 3M range for various tests. Can you add comments to explain why that is and how the allocations work? Maybe there should be a central comment somewhere and a ref to that comment from each of the relevant locations in code and scripts?

Comment (zuercher) [reply test/exe/main_common_test.cc:41]: So when runs-per-test is set and multiple tests are executed (e.g. `bazel test --runs_per_test=2 //test:foo_test //test:bar_test`), the same TEST_RANDOM_SEED gets reused across tests. Each instance of a particular test (say //test:foo_test) sees unique values, but those same seeds are used again for other tests (say //test:bar_test). If both use hot restart they interfere with each other. I added this range setting to make it impossible for these tests to collide.

I'll add a comment and after I figure out why the tests failed.

Comment (zuercher) [reply test/exe/main_common_test.cc]: Ok. Test is fixed and the base-id nonsense is better documented.

**Comment by zuercher** [comment]
Well it failed coverage due to low %-age.

File: test/exe/main_common_test.cc [hunk 3]
```diff
@@ -106,7 +81,7 @@ TEST_P(MainCommonTest, ConstructDestructHotRestartDisabled) {
 
 // Exercise init_only explicitly.
 TEST_P(MainCommonTest, ConstructDestructHotRestartDisabledNoInit) {
-  addArg("--disable-hot-restart");
+  //  addArg("--disable-hot-restart");
```
Comment (mattklein123) [inline test/exe/main_common_test.cc:84]: ?

File: test/test_common/environment.h [hunk 4]
```diff
@@ -72,6 +72,25 @@ class TestEnvironment {
...
+   * We require a unique base test id because random seeds are reused across
+   * multiple tests. For example, running:
+   *     bazel --runs_per_test=3 //test:foo_test //test:bar_test
+   * results in 3 unique seeds. Each set of tests is run once with each
+   * seed. Thus, the base test id is used to differentiate between tests run in
+   * parallel with the same seed.
+   *
+   * @param base_test_id a uint32_t, larger than 1000000, to uniquely identify a test
+   */
+  static std::string chooseBaseId(uint32_t base_test_id);
```
Comment (mattklein123) [inline test/test_common/environment.h:92]: It's still not clear to me what this call uses different bases in different places? Can you explain that or maybe just list here what the current allocations are?

Comment (zuercher) [reply test/test_common/environment.h]: Switched it to just take numbers and do the 1000000 thing internally. Hopefully that makes it better. (1000000 is just to give space for a lot of random test seeds or test shards).

Comment (zuercher) [reply test/test_common/environment.h]: Re-wrote the comments on the method as well.

File: test/test_common/environment.h [hunk 5]
```diff
@@ -72,6 +72,47 @@ class TestEnvironment {
...
+   * This method uses the given test base id and one of the following
+   * environment variables:
+   * - TEST_RANDOM_SEED is used to handle concurrent runs via the Bazel
+   *                    --runs_per_test flag
+   * - TEST_SHARD_INDEX is used to handle concurrent runs when tests are run in
+   *                    shards, such as in coverage testing
+   *
+   * This algorithm is re-implemented in the following test scripts:
+   * - test/integration/hot_restart_test.sh
+   * - test/integration/run_envoy_test.sh
```
Comment (mattklein123) [inline test/test_common/environment.h:105]: I shudder to even suggest this, but given how confusing/complicated this is already, should we just implement this "algorithm" as a `.sh` script and then just shell out to it when we need to from the CC file? Then all the logic is in one place with good comments. Up to you.

Comment (zuercher) [reply test/test_common/environment.h:105]: It still requires everyone to pick a unique set of test base ids. 

Sigh. I think the gold-plated solution is probably some special value for base-id that tells Envoy to pick an unused base id (like creating a socket on port 0) and a flag to write the resulting base id to a file. Most tests don't need to know the base id -- they just need to not collide with other envoy processes and the test that does need to know (hotrestart_test.sh) can get the base id and use it for its testing.

Comment (mattklein123) [reply test/test_common/environment.h:105]: Yeah this is all terrible. I'm fine with this for now given the comments as it is strictly better than what we have now, but up to you.

Comment (zuercher) [reply test/test_common/environment.h:105]: Oh, you know me. Can't resist a challenge. If you want to merge this I'll put up a PR with the better fix.

Comment (mattklein123) [reply test/test_common/environment.h:105]: Your call. :)

**Pull request merged**